In [ ]:
import polars as pl
import pandas as pd
import numpy as np
import bottleneck as bn

## Polars

Считаем датасет из файла `train.csv` (это данные о выживаемости на Титанике) с помощью polars

In [ ]:
df = pl.read_csv('train.csv')

Выведем информацию о датасете


In [ ]:
print(df.describe()) #Основная информация о датасете

print(df["Pclass"].value_counts()) #Посчитаем количество пассажиров каждого класса(Pclass)

print(df.group_by("Sex").agg((pl.col("Survived")).sum())) #Выведем количество выживших мужчин и женщин на корабле

print(df.filter(df["Age"] > 44)) #Выведем часть таблицы с пассажирами, возраст которых больше 44 лет

shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ statistic  ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   ┆ ---       ┆ ---       ┆ ---   ┆ ---      │
│ str        ┆ f64         ┆ f64      ┆ f64      ┆   ┆ str       ┆ f64       ┆ str   ┆ str      │
╞════════════╪═════════════╪══════════╪══════════╪═══╪═══════════╪═══════════╪═══════╪══════════╡
│ count      ┆ 891.0       ┆ 891.0    ┆ 891.0    ┆ … ┆ 891       ┆ 891.0     ┆ 204   ┆ 889      │
│ null_count ┆ 0.0         ┆ 0.0      ┆ 0.0      ┆ … ┆ 0         ┆ 0.0       ┆ 687   ┆ 2        │
│ mean       ┆ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ … ┆ null      ┆ 32.204208 ┆ null  ┆ null     │
│ std        ┆ 257.353842  ┆ 0.486592 ┆ 0.836071 ┆ … ┆ null      ┆ 49.693429 ┆ null  ┆ null     │
│ min        ┆ 1.0         ┆ 0.0      ┆ 1.0      ┆ … ┆ 110152    ┆ 0.0       ┆ A10   ┆ C        │
│ 25%

## Ускорение работы с pandas

Считаем датасет из файла `train.csv` (это данные о выживаемости на Титанике) с помощью pandas

In [ ]:
df = pd.read_csv('train.csv')

Выведем информацию согласно заданию

In [ ]:
#Посчитайте средний возраст пассажиров и его стандартное отклонение
print(f'Средний возраст пассажиров\t {bn.nanmean(df.Age)}')
print(f'Стандартное отклонение возраста\t {bn.nanstd(df.Age)}')

#Для каждого пассажира умножьте значение столбца Fare на 1.3, если класс его билета - 1 или 2, и на 1.1, если класс его билета - 3. Сохраните результаты как новый столбец Fare_new. (При реализации расчета используйте любые альтернативы методу iterrows)
np_arr = df[['Fare', 'Pclass']].to_numpy()
new_col = [row[0]*(1.3*(row[1] < 3) + 1.1*(row[1] == 3)) for row in np_arr]
df.insert(10, 'Fare_new', new_col, True) #добавим новую колонку в датасет
df

Средний возраст пассажиров	 29.69911764705882
Стандартное отклонение возраста	 14.516321150817317


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Fare_new,Fare_new,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,7.97500,7.97500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,92.66829,92.66829,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,8.71750,8.71750,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,69.03000,69.03000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,8.85500,8.85500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,16.90000,16.90000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,39.00000,39.00000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,25.79500,25.79500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,39.00000,39.00000,C148,C


## Оптимизация типов pandas

Считайте датасет из файла `Housing.csv` (это данные о ценах домов) с помощью pandas

In [ ]:
df = pd.read_csv('Housing.csv')

Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях

In [ ]:
df

,price,area,bedrooms,bathrooms,stories,mainroad,guestroom,basement,hotwaterheating,airconditioning,parking,prefarea,furnishingstatus
0,13300000,7420,4,2,3,yes,no,no,no,yes,2,yes,furnished
1,12250000,8960,4,4,4,yes,no,no,no,yes,3,no,furnished
2,12250000,9960,3,2,2,yes,no,yes,no,no,2,yes,semi-furnished
3,12215000,7500,4,2,2,yes,no,yes,no,yes,3,yes,furnished
4,11410000,7420,4,1,2,yes,yes,yes,no,yes,2,no,furnished
...,...,...,...,...,...,...,...,...,...,...,...,...,...
540,1820000,3000,2,1,1,yes,no,yes,no,no,2,no,unfurnished
541,1767150,2400,3,1,1,no,no,no,no,no,0,no,semi-furnished
542,1750000,3620,2,1,1,yes,no,no,no,no,0,no,unfurnished
543,1750000,2910,3,1,1,no,no,no,no,no,0,no,furnished


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [ ]:
df.describe()

,price,area,bedrooms,bathrooms,stories,parking
count,5.450000e+02,545.000000,545.000000,545.000000,545.000000,545.000000
mean,4.766729e+06,5150.541284,2.965138,1.286239,1.805505,0.693578
std,1.870440e+06,2170.141023,0.738064,0.502470,0.867492,0.861586
min,1.750000e+06,1650.000000,1.000000,1.000000,1.000000,0.000000
25%,3.430000e+06,3600.000000,2.000000,1.000000,1.000000,0.000000
50%,4.340000e+06,4600.000000,3.000000,1.000000,2.000000,0.000000
75%,5.740000e+06,6360.000000,3.000000,2.000000,2.000000,1.000000
max,1.330000e+07,16200.000000,6.000000,4.000000,4.000000,3.000000


**Определение оптимальных с точки зрения потребления памяти типов данных:**

1. Численные параметры не могут быть отрицательными, поэтому следует использовать unsigned типы данных.
2. Все строковые параметры (5-9 и 11-12 (индексы)) нужно категорировать
3. Параметры (1, 3, 4, 10) не превышают значения 6 , а значит следует использовать для них uint8.
4. Значение параметра 2(area) целочислены и принадлежат диапазону [1650;16200] , а значит для них следует использовать uint16.
5. С параметром 0(price) аналогично, но значения принадлежат диапазону [1.75e+6;1.33e+7] , следует использовать unit32.



Поменяйте типы данных столбцов датафрейма на выбранные вами в прошлом пункте и сравните потребление памяти до и после оптимизации

In [ ]:
df['price'] = df['price'].astype('uint32')
df['area'] = df['area'].astype('uint16')
df['bedrooms'] = df['bedrooms'].astype('uint8')
df['bathrooms'] = df['bathrooms'].astype('uint8')
df['stories'] = df['stories'].astype('uint8')
df['parking'] = df['parking'].astype('uint8')
df['mainroad'] = df['mainroad'].astype('category')
df['guestroom'] = df['guestroom'].astype('category')
df['basement'] = df['basement'].astype('category')
df['hotwaterheating'] = df['hotwaterheating'].astype('category')
df['airconditioning'] = df['airconditioning'].astype('category')
df['prefarea'] = df['prefarea'].astype('category')
df['furnishingstatus'] = df['furnishingstatus'].astype('category')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    uint32  
 1   area              545 non-null    uint16  
 2   bedrooms          545 non-null    uint8   
 3   bathrooms         545 non-null    uint8   
 4   stories           545 non-null    uint8   
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    uint8   
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), uint16(1), uint32(1), uint8(4)
memory usage: 10.0 KB


В результате изменения типов данных мы оптимизировали использование памяти в 5.5 раз.